align/schema/constraint.py

```diff
@@ -1,155 +1,257 @@
+            elif self.line == 'h_any':
+                checker.append(
+                    checker.Or( # We don't know which bbox is higher yet
+                        checker.And(b1.lly >= b2.lly, b1.ury <= b2.ury),
+                        checker.And(b2.lly >= b1.lly, b2.ury <= b1.ury)
+                    )
+                )
+            elif self.line == 'v_left':
+                checker.append(b1.llx == b2.llx)
+            elif self.line == 'v_right':
+                checker.append(b1.urx == b2.urx)
+            elif self.line == 'v_center':
+                checker.append(
+                    (b1.llx + b1.urx) / 2 == (b2.llx + b2.urx) / 2)
+            elif self.line == 'v_any':
+                checker.append(
+                    checker.Or( # We don't know which bbox is wider yet
+                        checker.And(b1.urx <= b2.urx, b1.llx >= b2.llx),
+                        checker.And(b2.urx <= b1.urx, b2.llx >= b1.llx)
+                    )
+                )
+            else:
+                checker.append(
+                    checker.Or( # h_any OR v_any
+                        checker.And(b1.urx <= b2.urx, b1.llx >= b2.llx),
+                        checker.And(b2.urx <= b1.urx, b2.llx >= b1.llx),
+                        checker.And(b1.lly >= b2.lly, b1.ury <= b2.ury),
+                        checker.And(b2.lly >= b1.lly, b2.ury <= b1.ury)
+                    )
+                )
 
-    def __init__(self, validation=None):
-        super().__init__(__root__=[])
-        self._commits = collections.OrderedDict()
-        if z3 is not None:
-            self._solver = z3.Solver()
-            if validation is not None:
-                self._validation = validation
-        else:
-            assert not validation, "Cannot validate without z3"
+# You may chain constraints together for more complex constraints by
+#     1) Assigning default values to certain attributes
+#     2) Using custom validators to modify attribute values
+# Note: Compositional check() is automatically constructed if
+#     every check() in mro starts with `constraints = super().check()`.
+#     (mro is Order, Align, ConstraintBase in this example)
+# Note: If you need to specialize check(), you do have the option 
+#     to create a custom `check()` in this class. It shouldn't be
+#     needed unless you are adding new semantics
+class AlignInOrder(Order, Align):
+    '''
+    Align `instances` in order, along the `direction`, on the `line`
+
+    > `direction == 'horizontal'` => left_to_right
+
+    > `direction == 'vertical'`   => bottom_to_top
+    '''
+    instances : List[str]
+    direction: Optional[Literal['horizontal', 'vertical']]
+    line : Optional[Literal[
+        'top', 'bottom',
+        'left', 'right',
+        'center'
+    ]]
+    abut: Optional[bool] = False
+
+    @types.root_validator(allow_reuse=True)
+    def _cast_constraints_to_base_types(cls, values):
```

Comment:
I would suggest making `horizontal` and `bottom` default. This would also simplify the ambiguity processing.

Reply:
Can you take care of this when working on the circuit examples @soneryaldiz ? I would keep the ambiguity processing if None is allowed (Optional)